cmake: allow redefinition BEEROCKS_* vars from cmdline

MuteSpirit opened the pull request:
If CMake variable defined in CMakeLists.txt file like
```
    set(VARIABLE VALUE)
```
Then it's impossible to redefine it from command line.
Event tricks like next does not work:
```
    cmake ... -UVARIABLE -DVARIABLE=NEW_VALUE
```
So Yocto/RDKB recipes and OpenWrt/prplWrt makefiles cannot
redefine variables without modification CMakeLists.txt files.

As a result conf files for controller/agent/etc. generated with values unusable for RDKB/etc.

Define variables interested for redefinition following pattern:
```
    if(NOT DEFINED VARIABLE)
        set(VARIABLE DEFAUL_VALUE)
    endif()
```

arnout commented on CMakeLists.txt:
```diff
@@ -89,8 +89,12 @@ if ("${isSystemDir}" STREQUAL "-1")
-set(TMP_PATH "/tmp/beerocks")
+if(NOT DEFINED INSTALL_PATH)
+    set(INSTALL_PATH "/opt/prplmesh")
+endif()
```
The normal way of allowing setting from the outside would be to make it a cache variable:
```suggestion
set(INSTALL_PATH "/opt/prplmesh" CACHE PATH "prplMesh installation directory")
```
However, I would much prefer to remove INSTALL_PATH entirely and instead fully rely on CMAKE_INSTALL_PREFIX.

MuteSpirit replied:
> However, I would much prefer to remove INSTALL_PATH entirely and instead fully rely on CMAKE_INSTALL_PREFIX.

There is special task for that #1253
I think that it's out of this PR scope

arnout commented on CMakeLists.txt:
```diff
@@ -89,8 +89,12 @@ if ("${isSystemDir}" STREQUAL "-1")
+endif()
+if(NOT DEFINED TMP_PATH)
+    set(TMP_PATH "/tmp/beerocks")
+endif()
```
Again, use a cache variable instead. However, why do you need to override this at all for RDKB?

MuteSpirit replied:
I rechecked that it's not needed to redefine this variable in recipe.
Reverted my changes for it.

arnout commented on CMakeLists.txt:
```diff
@@ -102,22 +106,44 @@ if(TARGET_PLATFORM STREQUAL "linux")
+endif()
+if(NOT DEFINED BEEROCKS_BRIDGE_IFACE)
+    set(BEEROCKS_BRIDGE_IFACE "br-lan")
+endif()
```
All this stuff really should be removed. Instead, we should parse it out of the configuration file, if appropriate. For prplmesh_utils.sh that's actually a little difficult... So OK for now.

MuteSpirit replied:
I used cached variable for them too to have 4 lines instead of 12 as here

arnout commented on CMakeLists.txt:
```diff
@@ -102,22 +106,44 @@ if(TARGET_PLATFORM STREQUAL "linux")
-set(BEEROCKS_REPEATER_MODE 0)
+if (NOT DEFINED BEEROCKS_REPEATER_MODE)
+    set(BEEROCKS_REPEATER_MODE 0)
+endif()
```
Do you really need to change this for the RDKB build? Please only make this modification for stuff that really needs to be changed from the outside...

MuteSpirit replied:
No
Reverted changes for it

arnout commented on common/CMakeLists.txt:
```diff
@@ -1,7 +1,9 @@
-set(BEEROCKS_PLATFORM_INIT "false")
+if (NOT DEFINED BEEROCKS_PLATFORM_INIT)
+    set(BEEROCKS_PLATFORM_INIT "false")
+endif()
```
This one really should be removed entirely, as I note in your other PR.

arnout commented on common/beerocks/bwl/CMakeLists.txt:
```diff
@@ -10,7 +10,9 @@ file(GLOB_RECURSE bwl_common_sources ${MODULE_PATH}/common/*.c*)
-set(BWL_TYPE "DUMMY")
+if (NOT DEFINED BWL_TYPE)
+    set(BWL_TYPE "DUMMY")
+endif()
```
Cache variable please.

arnout commented on framework/CMakeLists.txt:
```diff
@@ -2,8 +2,10 @@ project(multiap_framework C CXX)
+if (NOT DEFINED MSGLIB)
+    set(MSGLIB "zmq" CACHE STRING "Which messaging library backend to use")
+    set_property(CACHE MSGLIB PROPERTY STRINGS "zmq" "nng" "None")
+endif()
```
Here you're converting the correct thing into something wrong...

MuteSpirit replied:
Agree
Reverted my changes here

arnout commented on CMakeLists.txt:
```diff
@@ -102,22 +106,44 @@ if(TARGET_PLATFORM STREQUAL "linux")
+endif()
+if(NOT DEFINED BEEROCKS_WLAN3_IFACE)
+    set(BEEROCKS_WLAN3_IFACE "wlan4")
+endif()
```
This one is actually not used, please remove it instead.

MuteSpirit replied:
Should i also remove next code from `agent/config/beerocks_agent.conf.in`?
```

[agent2]
radio_identifier=00:00:00:00:00:02
hostap_iface_type=WIFI_INTEL
hostap_ant_gain=0
enable_repeater_mode=0
```

arnout replied:
No, we will add a 3-radio platform at some point so that will still be useful. However, we'll probably won't have a docker test for it so the CMake option shouldn't be needed.

arnout commented on CMakeLists.txt:
```diff
@@ -7,7 +7,7 @@ project(prplmesh
 
 if(NOT DEFINED TARGET_PLATFORM)
-set(TARGET_PLATFORM "linux")
+    set(TARGET_PLATFORM "linux")
```
Please do the whitespace-only changes in a separate commit.

arnout reviewed (changes requested):
Overall, four types of changes requested:
- Do the whitespace-only changes in a separate commit.
- Remove variables that are not used (in a separate commit).
- Use cache variables instead of the "if not" pattern.
- Don't change it for variables that you don't need to override for RDKB.

For the variables that really should be removed entirely (all the configuration stuff), it's OK to keep the "if not" pattern, since it's only temporary anyway.

Commit:
CMakeLists.txt: fix identation in 'if' statements

There are some places in ./CMakeLists.txt file like:

if(...
somecodehere...
endif()

Added identation for code inside "if" branch for better readeness

if(...
    somecodehere...
endif()

Signed-off-by: Ivan Efimov <i.efimov@inango-systems.com>

Commit:
cmake: allow redefinition BEEROCKS_* vars from cmdline

If CMake variable defined in CMakeLists.txt file like

    set(VARIABLE VALUE)

Then it's impossible to redefine it from command line.
Event tricks like next does not work:

    cmake ... -UVARIABLE -DVARIABLE=NEW_VALUE

As a result Yocto/RDKB recipes and OpenWrt/prplWrt makefiles cannot
redefine variables without modification CMakeLists.txt files.

Define variables interested for redefinition following pattern:

    set(VARIABLE DEFAUL_VALUE CACHE <type> <docstring>)

Signed-off-by: Ivan Efimov <i.efimov@inango-systems.com>

rmelotte commented on common/CMakeLists.txt:
```diff
@@ -1,7 +1,7 @@
 
 # Default values
-set(BEEROCKS_PLATFORM_INIT "false")
+set(BEEROCKS_PLATFORM_INIT "false" CACHE STRING "Configure minimal set of dummy WLAN ifaces and bridge")
```
This variable is only used in the prplmesh_utils.sh script.
There, it can be controlled using the "--platform-init" option. Why would we additionally need to set the default to true?
Also, with this change and if BEEROCKS_PLATFORM_INIT is set to true, the "--platform-init" option from the script becomes unusable, which can be confusing.

arnout replied:
I also think it should be removed entirely, but as Ivan wrote somewhere else: that's not the purpose of this PR. As it is now, the PR nicely converts all existing options that should be configurable to cache variables, so I'm happy with it.

arnout commented on common/beerocks/bwl/CMakeLists.txt:
```diff
@@ -10,7 +10,8 @@ file(GLOB_RECURSE bwl_common_sources ${MODULE_PATH}/common/*.c*)
 
 # Default defaults
-set(BWL_TYPE "DUMMY")
+set(BWL_TYPE "DUMMY" CACHE STRING "BWL type")
```
It is already defined as a cache variable below. So that one should be removed. 

However, I still don't think it's OK then. IIUC a plain variable overrides a cache variable, and both can be defined at the same time. So the `set` commands within the `TARGET_PLATFORM` condition below will override whatever was set in the cache variable. I can't say that CMake is entirely clear for me though, so I could be wrong about this.

Even so, I think it would be much clearer to do what I suggested before: change the variable name to `BWL_TYPE_DEFAULT` and use that in the definition of the cache variable below.

arnout commented on common/beerocks/bwl/CMakeLists.txt:
```diff
@@ -10,7 +10,8 @@ file(GLOB_RECURSE bwl_common_sources ${MODULE_PATH}/common/*.c*)
 # Default defaults
-set(BWL_TYPE "DUMMY")
+set(BWL_TYPE "DUMMY" CACHE STRING "BWL type")
+set_property(CACHE BWL_TYPE PROPERTY STRINGS "DUMMY" "NL80211" "DWPAL")
```
Nice!

arnout reviewed (approved):
If it works in RDKB with the current definition of BWL_TYPE, I'm OK to merge it as is.

Really nice commit messages by the way, thanks!
